### PR TITLE
[clang] Add cc1 --output-asm-variant= to set output syntax

### LINKED ISSUE #109157: Since Clang 19 `-mllvm --x86-asm-syntax=intel` seems to affect inline assembly syntax in source
Hello,

I personally use Clang to translate x86 assembly between AT&T syntax and Intel syntax. Clang 18- can compile AT&T syntax to Intel syntax, as in

```
$ echo '__asm__("mov $1, %eax");'  \
  | clang -x c -  -mllvm --x86-asm-syntax=intel -O2 -S -o -  \
  | grep mov
        mov     eax, 1
```

but Clang 19 fails to parse the source. It seems to expect Intel syntax:

```
$ echo '__asm__("mov $1, %eax");' |  \
  | ./llvm-mingw-20240917-ucrt-x86_64/bin/clang -x c -  -mllvm --x86-asm-syntax=intel -O2 -S -o -  \
  | grep mov
<inline asm>:1:5: error: unknown token in expression
```

@MaskRay Do you have any idea whether this has something to do with #85367 ?

### PR DESCRIPTION
**Pull request by MaskRay**
2fcaa549a824efeb56e807fcf750a56bf985296b (2010) added cc1as option
`-output-asm-variant` (untested) to set the output syntax.
`clang -cc1as -filetype asm -output-asm-variant 1` allows AT&T input and
Intel output (`AssemblerDialect` is also used by non-x86 targets).

This patch renames the cc1as option (to avoid collision with -o) and
makes it available for cc1 to set output syntax. This allows different
input & output syntax:

```
echo 'asm("mov $1, %eax");' | clang -xc - -S -o - -Xclang --output-asm-variant=1
```

Note: `AsmWriterFlavor` (with a misleading name), used to initialize
MCAsmInfo::AssemblerDialect, is primarily used for assembly input, not
for output.
Therefore,
`echo 'asm("mov $1, %eax");' | clang -x c - -mllvm --x86-asm-syntax=intel -S -o -`,
which achieves a similar goal before Clang 19, was unintended.

Close #109157

### CODE REVIEW

**Comment by lhmouse** [comment]
Thanks. The idea looks awesome!

File: clang/include/clang/Driver/Options.td [hunk 1]
```diff
@@ -7209,6 +7209,9 @@ def fuse_ctor_homing: Flag<["-"], "fuse-ctor-homing">,
 def as_secure_log_file : Separate<["-"], "as-secure-log-file">,
   HelpText<"Emit .secure_log_unique directives to this filename.">,
   MarshallingInfoString<CodeGenOpts<"AsSecureLogFile">>;
+def output_asm_variant : Joined<["--"], "output-asm-variant=">,
+  HelpText<"Select the asm variant (integer) to use for output">,
```
Comment (mstorsjo) [inline clang/include/clang/Driver/Options.td:7213]: Requiring the user to pick a variant via an integer index feels a bit user-unfriendly - should/could we expose this with names like `=intel` and `=att` like we do for the other related options? Admittedly, this is not a very user facing option, as it is a`-cc1` option only, if I'm correctly understanding the situation, but nevertheless, I wouldn't offhand have any idea about what indices to use here.

**Review by mstorsjo — APPROVED** [review]
Aside for the option ergonomics, this looks reasonable to me.

File: clang/include/clang/Driver/Options.td (hunk 1 above)
Comment (MaskRay) [reply clang/include/clang/Driver/Options.td:7213]: Agreed with the conservation. The ergonomics is one the reasons that this option is kept internal (cl::opt, cc1, not driver).

llvm-mc and llc currently use output-asm-variant=1 and the integer looks better for non-x86 targets.

Comment (lhmouse) [reply clang/include/clang/Driver/Options.td:7213]: For a GNU inline asm statement, it's possible to specify multiple dialects, like in

```
__asm__ ("{ movl $1, %eax | mov eax, 1 }");
```

so index zero is AT&T and index one is Intel, hard-coded in source code like that.

A numeric argument is probably not too bad if it matches another tool.

Comment (MaskRay) [reply clang/include/clang/Driver/Options.td:7213]: Thanks for the GCC information. Looks good that the numerical index, already used by existing internal LLVM tools, also matches GCC internals.

Comment (MaskRay) [reply clang/include/clang/Driver/Options.td]: Appreciate the comments. 

Yes, the AssemblerDIalect choices could be useful for other targets. SystemZ has `AssemblerDIalect`s that I am not familiar with. AArch64 seems to have one as well.

Comment (mstorsjo) [reply clang/include/clang/Driver/Options.td]: Oh, indeed - yes, there's a nonstandard dialect for NEON instructions in aarch64, which is used for disassembly on Darwin targets by default. Thanks for the pointers!

**Pull request merged**